src/lib/testing/property.ts

```diff
@@ -1,40 +1,55 @@
-import { Random, withHardCoded } from './random.js';
-export { test, Random, withHardCoded };
+import { Random } from './random.js';
+export { test };
+export { Random, sample, withHardCoded } from './random.js';
 
 const defaultTimeBudget = 100; // ms
-const minRuns = 15;
-const maxRuns = 100;
-const isVerbose = false;
+const defaultMinRuns = 15;
+const defaultMaxRuns = 400;
 
-function test<T extends readonly Random<any>[]>(...args: ArrayTestArgs<T>) {
-  let timeBudget = defaultTimeBudget;
-  let run: (...args: ArrayRunArgs<Nexts<T>>) => void;
-  let arg = args.pop();
-  if (typeof arg !== 'function') {
-    if (arg !== undefined) timeBudget = (arg as any).timeBudget;
-    run = args.pop() as any;
-  } else {
-    run = arg;
-  }
-  let gens = args as any as T;
-  let nexts = gens.map((g) => g.create()) as Nexts<T>;
-  let start = performance.now();
-  // run at least `minRuns` times
-  testN(minRuns, nexts, run);
-  let time = performance.now() - start;
-  if (time > timeBudget) return minRuns;
-  // (minRuns + remainingRuns) * timePerRun = timeBudget
-  let remainingRuns = Math.floor(timeBudget / (time / minRuns)) - minRuns;
-  // run at most `maxRuns` times
-  if (remainingRuns > maxRuns - minRuns) remainingRuns = maxRuns - minRuns;
-  testN(remainingRuns, nexts, run);
-  return minRuns + remainingRuns;
+const test = Object.assign(testCustom(), {
+  negative: testCustom({ negative: true }),
+  custom: testCustom,
+});
```

Comment:
adds a primitive for negative testing

Reply:
Are these "random generation" tests deterministic/reproducible?

Reply:
no, they aren't. I think this would come at two significant costs which IMO make it not worth pursuing:
- large engineering costs of refactoring to RNGs that take a seed
- significant cost on sample diversity. some tests are so slow that they are only run with 15 samples - the minimum. if the samples are different every time, we get diversity by running this test many times (which we do, since it runs in CI all the time). By making it deterministic, we would lose that
- In practice, bugs that occur with a few % probability are exposed in almost every test run. Concretely, for all the bugs that I uncovered when testing this, I could reproduce these bugs in _every_ test run, which made them easy to find despite the test cases not being the same every time

Reply:
Also I should add that the testing framework prints out the failing samples when a test fails! This enables you to debug the failure no matter how elusive it may be

Reply:
My second option was that it dumps all the details required to reproduce the test.  This satisfies my requirement on reproducibility. 👍🏼